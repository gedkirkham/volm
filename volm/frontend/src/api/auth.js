import Vue from 'vue'

export function loginApi (OBJ) {
    Vue
        .http
        .post(process.env.API + '/token/', {
            username: OBJ.email,
            password: OBJ.password,
        })
        .then(RESP => {
            const TOKENS = RESP.body
            localStorage.setItem('access', TOKENS.access)
            localStorage.setItem('refresh', TOKENS.refresh)
        })
        .catch(ERROR => console.error('login()', ERROR))
}

let registerApiPromise
let registerApiLoading = false
export function registerApi (OBJ) {
    if (registerApiLoading) return registerApiPromise

    registerApiLoading = true
    registerApiPromise = Vue
        .http
        .post(process.env.API + '/register/', {
            email: OBJ.email,
            first_name: OBJ.first_name,
            last_name: OBJ.last_name,
            password: OBJ.password,
            username: OBJ.email,
        })
        .finally(() => { registerApiLoading = false })

    return registerApiPromise
}
