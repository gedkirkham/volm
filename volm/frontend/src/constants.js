export default {
    django: {
        errors: {
            blank: 'This field may not be blank.',
            max_30_char: 'Ensure this field has no more than 30 characters.',
            max_128_char: 'Ensure this field has no more than 128 characters.',
            max_150_char: 'Ensure this field has no more than 150 characters.',
            invalid_email: 'Enter a valid email address.',
            password_mismatch: 'Passwords do not match.',
            password_min_8_char: 'This password is too short. It must contain at least 8 characters.',
            password_too_common: 'This password is too common.',
            password_entirely_numeric: 'This password is entirely numeric.',
            username_already_exists: 'A user with that username already exists.',
        },
    },
}
