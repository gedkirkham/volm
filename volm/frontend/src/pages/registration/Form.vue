<template>
    <q-page class="flex flex-center">
        <div class="column items-start">
            <section>
                <header>
                    <h1 class="text-h2 q-mb-none">
                        Register
                    </h1>
                </header>
                <q-form
                    class="q-mt-xl"
                    :style="!$q.screen.lt.sm ? 'min-width:400px;' : 'min-width:90vw;'"
                    @submit="register()"
                >
                    <span date-test="firstName">
                        <q-input
                            v-model="first_name"
                            class="q-pb-xl"
                            for="test-first_name"
                            id="test-first_name"
                            label="First name *"
                            outlined
                            :error="!!errors.first_name"
                            :error-message="errors.first_name"
                        />
                    </span>
                    <span data-test="lastName">
                        <q-input
                            v-model="last_name"
                            class="q-pb-xl"
                            for="test-last_name"
                            id="test-last_name"
                            label="Last name *"
                            outlined
                            :error="!!errors.last_name"
                            :error-message="errors.last_name"
                        />
                    </span>
                    <span data-test="email">
                        <q-input
                            v-model="email"
                            class="q-pb-xl"
                            for="test-email"
                            id="test-email"
                            label="Email *"
                            outlined
                            :error="!!errors.email"
                            :error-message="errors.email"
                        />
                    </span>
                    <span data-test="password_1">
                        <q-input
                            v-model="password_1"
                            class="q-pb-xl"
                            for="test-password_1"
                            id="test-password_1"
                            label="Password *"
                            type="password"
                            outlined
                            :error="!!errors.password"
                            :error-message="errors.password"
                        />
                    </span>
                    <span data-test="password_2">
                        <q-input
                            v-model="password_2"
                            class="q-pb-xl"
                            for="test-password_2"
                            id="test-password_2"
                            label="Confirm password *"
                            type="password"
                            outlined
                            :error="!!errors.password"
                            :error-message="errors.password"
                        />
                    </span>

                    <q-btn
                        color="primary"
                        data-test="submit"
                        label="Submit"
                        type="submit"
                        :loading="isLoading"
                    />
                </q-form>

                <div class="q-my-lg">
                    Already have an account?
                    <router-link
                        date-test="log-in"
                        :to="{ name: 'login' }"
                    >
                        Log-in
                    </router-link>
                </div>
            </section>
        </div>
    </q-page>
</template>

<script>
import { registerApi } from 'src/api/auth.js'
import constants from 'src/constants.js'

export default {
    name: 'RegistrationPage',
    data () {
        return {
            errors: {},
            email: '',
            first_name: '',
            isLoading: false,
            last_name: '',
            password_1: '',
            password_2: '',
        }
    },
    methods: {
        getErrorMessage (DJANGO_ERROR) {
            const OBJ = {
                [constants.django.errors.blank]: this.$t('pages.register.errors.blank'),
                [constants.django.errors.max_30_char]: this.$t('pages.register.errors.max_30_char'),
                [constants.django.errors.max_128_char]: this.$t('pages.register.errors.max_128_char'),
                [constants.django.errors.max_150_char]: this.$t('pages.register.errors.max_150_char'),
                [constants.django.errors.invalid_email]: this.$t('pages.register.errors.invalid_email'),
                [constants.django.errors.password_mismatch]: this.$t('pages.register.errors.password_mismatch'),
                [constants.django.errors.password_min_8_char]: this.$t('pages.register.errors.password_min_8_char'),
                [constants.django.errors.password_too_common]: this.$t('pages.register.errors.password_too_common'),
                [constants.django.errors.password_entirely_numeric]: this.$t('pages.register.errors.password_entirely_numeric'),
                [constants.django.errors.username_already_exists]: this.$t('pages.register.errors.email_already_exists'),
                default: this.$t('pages.register.errors.field_contains_errors'),
            }

            return (OBJ[DJANGO_ERROR] || OBJ.default)
        },
        register () {
            this.errors = {}
            this.isLoading = true
            registerApi({
                email: this.email,
                first_name: this.first_name,
                last_name: this.last_name,
                password_1: this.password_1,
                password_2: this.password_2,
            })
                .then(() => {
                    this.$router.push({ name: 'confirm_email' })
                })
                .catch(ERROR => {
                    console.error('registerApi()', ERROR)
                    this.setErrors(ERROR.body)
                })
                .finally(() => { this.isLoading = false })
        },
        setErrors (ERRORS) {
            Object.entries(ERRORS).forEach(
                ([KEY, VALUE]) => {
                    const ERROR_MESSAGES = VALUE
                    const UPDATED_KEY = KEY === 'username' ? 'email' : KEY
                    ERROR_MESSAGES.forEach(ERROR => {
                        let error = ''
                        if (this.errors[UPDATED_KEY]) error = `${this.errors[UPDATED_KEY]}. ${this.getErrorMessage(ERROR)}`
                        else error = `${this.getErrorMessage(ERROR)}`
                        this.$set(this.errors, UPDATED_KEY, error)
                    })
                },
            )
        },
    },
}
</script>
