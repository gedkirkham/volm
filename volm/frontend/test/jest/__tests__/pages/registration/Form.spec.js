/* eslint-disable */
/**
 * @jest-environment jsdom
 */

import {
    createLocalVue,
    enableAutoDestroy,
} from '@vue/test-utils'
import Form from 'src/pages/registration/Form.vue'
import * as All from 'quasar'
import VueRouter from 'vue-router'
import VueI18n from 'vue-i18n'
import {
    createShallowWrapper,
    createMountedWrapper,
} from 'jest/utils/index.js'
// import langEn from 'quasar/lang/en-us' // change to any language you wish! => this breaks wallaby :(
const { Quasar, date } = All

import constants from 'src/constants.js'
import messages from 'src/i18n'

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})

enableAutoDestroy(afterEach)


import { registerApi } from 'src/api/auth.js'

jest.mock('src/api/auth.js', () => ({
    registerApi: jest.fn()
}))

describe('Registration From', () => {
    let mountedWrapper
    let shallowWrapper

    const localVue = createLocalVue()
    localVue.use(Quasar, { components }) // , lang: langEn
    localVue.use(VueRouter)
    localVue.use(VueI18n)

    const i18n = new VueI18n({
        locale: 'en-us',
        fallbackLocale: 'en-us',
        messages,
    })

    const PROVIDE = {
        layout: {
            footer: {},
            header: {},
            left: {},
            right: {},
        },
        pageContainer: true,
    }

    beforeEach(() => {
        shallowWrapper = createShallowWrapper({
            component: Form,
            i18n,
            localVue,
            options: {
                mocks: {
                    $q: {
                        screen: {
                            lt: {
                                sm: jest.mock(),
                            },
                        },
                    },
                },
            },
        })

        mountedWrapper = createMountedWrapper({
            component: Form,
            i18n,
            localVue,
            options: {
                provide: PROVIDE,
            },
        })
    })

    it('passes the sanity check and creates a wrapper', () => {
        expect(shallowWrapper.exists()).toBe(true)
    })

    it('renders correctly', () => {
        expect(shallowWrapper).toMatchSnapshot()
    })

    describe('methods', () => {
        describe('getErrorMessage()', () => {
            it('is a function', () => {
                expect(typeof shallowWrapper.vm.getErrorMessage).toBe('function')
            })

            it('returns correct error message', () => {
                const ERROR = shallowWrapper.vm.getErrorMessage(constants.django.errors.blank)
                expect(ERROR).toBe(i18n.t('pages.register.errors.blank'))
            })

            it('returns default error message', () => {
                const ERROR = shallowWrapper.vm.getErrorMessage('random error message')
                expect(ERROR).toBe(i18n.t('pages.register.errors.field_contains_errors'))
            })
        })

        describe('register()', () => {
            it('is a function', () => {
                expect(typeof shallowWrapper.vm.register).toBe('function')
            })

            it('is called when the form is submitted', async () => {
                mountedWrapper.vm.register = jest.fn()
                expect(mountedWrapper.vm.register).toHaveBeenCalledTimes(0)
                await mountedWrapper.get('form').trigger('submit')
                expect(mountedWrapper.vm.register).toHaveBeenCalledTimes(1)
            })

            it('sets "isLoading" to true', async () => {
                registerApi.mockResolvedValueOnce()
                expect(shallowWrapper.vm.isLoading).toBe(false)
                shallowWrapper.vm.register()
                expect(shallowWrapper.vm.isLoading).toBe(true)
            })

            it('sets "errors" to empty object', async () => {
                registerApi.mockResolvedValueOnce()
                const ERROR = 'Error message'
                await shallowWrapper.setData({
                    errors: {
                        first_name: ERROR,
                    },
                })
                expect(shallowWrapper.vm.errors).toStrictEqual({
                    first_name: ERROR,
                })
                shallowWrapper.vm.register()
                expect(shallowWrapper.vm.errors).toStrictEqual({})
            })

            it('calls registerApi with the correct params', () => {
                registerApi.mockResolvedValueOnce()
                shallowWrapper.vm.register()
                expect(registerApi).toBeCalledWith({
                    email: '',
                    first_name: '',
                    last_name: '',
                    password_1: '',
                    password_2: '',
                })
            })

            describe('registerApi', () => {
                it('upon success, pushes new route', async () => {
                    registerApi.mockResolvedValue()
                    shallowWrapper.vm.$router.push = jest.fn()
                    expect(shallowWrapper.vm.$router.push).toHaveBeenCalledTimes(0)
                    await shallowWrapper.vm.register()
                    expect(shallowWrapper.vm.$router.push).toHaveBeenCalledTimes(1)
                    expect(shallowWrapper.vm.$router.push).toHaveBeenCalledWith({ name: 'confirm_email' })
                })

                it('upon failure, calls "setErrors" with correct params', async () => {
                    global.console = { error: jest.fn() }
                    shallowWrapper.vm.setErrors = jest.fn()
                    registerApi.mockRejectedValue({ body: 'Error message' })
                    expect(shallowWrapper.vm.setErrors).toBeCalledTimes(0)
                    await shallowWrapper.vm.register()
                    expect(shallowWrapper.vm.setErrors).toBeCalledTimes(1)
                    expect(console.error).toBeCalledTimes(1)
                })

                it('upon response, sets "isLoading" to false', async () => {
                    registerApi.mockResolvedValue()
                    expect(shallowWrapper.vm.isLoading).toBe(false)
                    await shallowWrapper.vm.register()
                    expect(shallowWrapper.vm.isLoading).toBe(false)
                })
            })
        })

        describe('setErrors()', () => {
            it('is a function', () => {
                expect(typeof shallowWrapper.vm.setErrors).toBe('function')
            })

            it('can set a single error and the component state updates correctly', () => {
                expect(shallowWrapper.vm.errors).toStrictEqual({})
                const ERROR = {
                    'first_name': [constants.django.errors.blank],
                }
                shallowWrapper.vm.setErrors(ERROR)
                expect(shallowWrapper.vm.errors).toStrictEqual({
                    'first_name': i18n.t('pages.register.errors.blank'),
                })
            })

            it('can set a multiple errors', () => {
                expect(shallowWrapper.vm.errors).toStrictEqual({})
                const ERRORS = {
                    'first_name': [constants.django.errors.blank, constants.django.errors.max_30_char],
                }
                shallowWrapper.vm.setErrors(ERRORS)
                expect(shallowWrapper.vm.errors).toStrictEqual({
                    'first_name': `${i18n.t('pages.register.errors.blank')}. ${i18n.t('pages.register.errors.max_30_char')}`,
                })
            })

            it('errors are displayed correctly', async () => {
                await mountedWrapper.setData({
                    errors: {
                        first_name: 'This is my test error!',
                    },
                })
                const element = mountedWrapper.find('[data-test=firstName]')
                expect(element).toMatchSnapshot()
            })

            it('"username" key is translated to "email"', () => {
                expect(shallowWrapper.vm.errors).toStrictEqual({})
                const ERROR = {
                    'username': [constants.django.errors.blank],
                }
                shallowWrapper.vm.setErrors(ERROR)
                expect(shallowWrapper.vm.errors).toStrictEqual({
                    'email': i18n.t('pages.register.errors.blank'),
                })
            })
        })
    })

    describe('state correctly updates when user enters data for', () => {
        beforeEach(() => {
            expect(mountedWrapper.vm.first_name).toBe('')
            expect(mountedWrapper.vm.last_name).toBe('')
            expect(mountedWrapper.vm.email).toBe('')
            expect(mountedWrapper.vm.password_1).toBe('')
            expect(mountedWrapper.vm.password_2).toBe('')
        })

        it('first name form field', () => {
            const VALUE = 'John'
            const INPUT_FIELD = mountedWrapper.find('[data-test=firstName] input')
            INPUT_FIELD.setValue(VALUE)
            expect(mountedWrapper.vm.first_name).toBe(VALUE)
        })

        it('last name form field', () => {
            const VALUE = 'Moore'
            const INPUT_FIELD = mountedWrapper.find('[data-test=lastName] input')
            INPUT_FIELD.setValue(VALUE)
            expect(mountedWrapper.vm.last_name).toBe(VALUE)
        })

        it('email form field', () => {
            const VALUE = 'test@test.com'
            const INPUT_FIELD = mountedWrapper.find('[data-test=email] input')
            INPUT_FIELD.setValue(VALUE)
            expect(mountedWrapper.vm.email).toBe(VALUE)
        })

        it('password form field', () => {
            const VALUE = 'password123'
            const INPUT_FIELD = mountedWrapper.find('[data-test=password_1] input')
            INPUT_FIELD.setValue(VALUE)
            expect(mountedWrapper.vm.password_1).toBe(VALUE)
        })

        it('confirm password form field', () => {
            const VALUE = 'password1234'
            const INPUT_FIELD = mountedWrapper.find('[data-test=password_2] input')
            INPUT_FIELD.setValue(VALUE)
            expect(mountedWrapper.vm.password_2).toBe(VALUE)
        })
    })

    describe('constants exist', () => {
        it('constants.django.errors.blank', () => {
            expect(constants.django.errors.blank).toBeTruthy()
        })

        it('constants.django.errors.max_30_char', () => {
            expect(constants.django.errors.max_30_char).toBeTruthy()
        })

        it('constants.django.errors.max_128_char', () => {
            expect(constants.django.errors.max_128_char).toBeTruthy()
        })

        it('constants.django.errors.max_150_char', () => {
            expect(constants.django.errors.max_150_char).toBeTruthy()
        })

        it('constants.django.errors.invalid_email', () => {
            expect(constants.django.errors.invalid_email).toBeTruthy()
        })

        it('constants.django.errors.password_mismatch', () => {
            expect(constants.django.errors.password_mismatch).toBeTruthy()
        })

        it('constants.django.errors.password_min_8_char', () => {
            expect(constants.django.errors.password_min_8_char).toBeTruthy()
        })

        it('constants.django.errors.password_too_common', () => {
            expect(constants.django.errors.password_too_common).toBeTruthy()
        })

        it('constants.django.errors.password_entirely_numeric', () => {
            expect(constants.django.errors.password_entirely_numeric).toBeTruthy()
        })

        it('constants.django.errors.username_already_exists', () => {
            expect(constants.django.errors.username_already_exists).toBeTruthy()
        })
    })

    describe('i18 values exist', () => {
        it('pages.register.errors.blank', () => {
            expect(messages['en-us'].pages.register.errors.blank).not.toBeUndefined()
        })

        it('pages.register.errors.max_30_char', () => {
            expect(messages['en-us'].pages.register.errors.max_30_char).not.toBeUndefined()
        })

        it('pages.register.errors.max_128_char', () => {
            expect(messages['en-us'].pages.register.errors.max_128_char).not.toBeUndefined()
        })

        it('pages.register.errors.max_150_char', () => {
            expect(messages['en-us'].pages.register.errors.max_150_char).not.toBeUndefined()
        })

        it('pages.register.errors.invalid_email', () => {
            expect(messages['en-us'].pages.register.errors.invalid_email).not.toBeUndefined()
        })

        it('pages.register.errors.password_mismatch', () => {
            expect(messages['en-us'].pages.register.errors.password_mismatch).not.toBeUndefined()
        })

        it('pages.register.errors.password_min_8_char', () => {
            expect(messages['en-us'].pages.register.errors.password_min_8_char).not.toBeUndefined()
        })

        it('pages.register.errors.password_too_common', () => {
            expect(messages['en-us'].pages.register.errors.password_too_common).not.toBeUndefined()
        })

        it('pages.register.errors.password_entirely_numeric', () => {
            expect(messages['en-us'].pages.register.errors.password_entirely_numeric).not.toBeUndefined()
        })

        it('pages.register.errors.username_already_exists', () => {
            expect(messages['en-us'].pages.register.errors.email_already_exists).not.toBeUndefined()
        })
    })
})
