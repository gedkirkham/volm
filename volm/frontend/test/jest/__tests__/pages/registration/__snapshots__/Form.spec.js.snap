// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration From renders correctly 1`] = `
<q-page-stub class="flex flex-center">
  <div class="column items-start">
    <section>
      <header>
        <h1 class="text-h2 q-mb-none">
          Register
        </h1>
      </header>
      <q-form-stub class="q-mt-xl" style="min-width: 90vw;"><span data-cy="test-firstName"><q-input-stub value="" label="First name *" outlined="true" type="text" class="q-pb-xl"></q-input-stub></span> <span data-cy="test-lastName"><q-input-stub value="" label="Last name *" outlined="true" type="text" class="q-pb-xl"></q-input-stub></span> <span data-cy="test-email"><q-input-stub value="" label="Email *" outlined="true" type="text" class="q-pb-xl"></q-input-stub></span> <span data-cy="test-password_1"><q-input-stub value="" label="Password *" outlined="true" type="password" class="q-pb-xl"></q-input-stub></span> <span data-cy="test-password_2"><q-input-stub value="" label="Confirm password *" outlined="true" type="password" class="q-pb-xl"></q-input-stub></span>
        <q-btn-stub ripple="true" align="center" type="submit" label="Submit" color="primary" data-cy="test-submit"></q-btn-stub>
      </q-form-stub>
      <div class="q-my-lg">
        Already have an account?
        <router-link-stub to="[object Object]" tag="a" ariacurrentvalue="page" event="click" data-cy="log-in">
          Log-in
        </router-link-stub>
      </div>
    </section>
  </div>
</q-page-stub>
`;
